### ISSUE TITLE
CUDA: sycl::native::sqrt results in slower PTX than sycl::sqrt

### ISSUE THREAD

**Issue by al42and** [AdaptiveCpp/AdaptiveCpp#478]
**Bug summary**
My understanding is that `sycl::native::` functions are supposed to be faster, less precise versions of the "default" `sycl::` math functions. However, it seems that (at least with clang-11 and CUDA-11.x), using `sycl::sqrt` generates faster code.

Namely, `cl::sycl::sqrt(x)` generates `sqrt.approx.f32` PTX instruction, while `cl::sycl::native::sqrt(x)` generates `sqrt.rn.f32`. It happens even with `-O0` flag.

I'm a bit too lazy to do the benchmarking myself, but https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#floating-point-instructions-sqrt says that `sqrt.approx.f32` is a fast approximation, while `sqrt.rn.f32` is IEEE754-compliant, presumably slower, version.

**To Reproduce**

https://gist.github.com/al42and/f85e5e11a3cd6cf448a4e718f2489f73

```bash
# please note `rm` below before copying
# Compile the code using `sycl::sqrt` function:
rm -rf /tmp/hipSYCL-syclcc-* && syclcc -O3 --hipsycl-explicit-multipass test_sqrt.cpp --hipsycl-targets='cuda:sm_35,sm_75' -o test --hipsycl-save-temps && grep 'sqrt\.[a-z]\+\.f32' /tmp/hipSYCL-*/nvptx-*.hipsycl_kernels
# It produces sqrt.approx.f32
# Compile the code using `sycl::native::sqrt` function:
rm -rf /tmp/hipSYCL-syclcc-* && syclcc -DUSE_NATIVE -Wno-unknown-cuda-version -O3 --hipsycl-explicit-multipass test_sqrt.cpp --hipsycl-targets='cuda:sm_35,sm_75' -o test --hipsycl-save-temps && grep 'sqrt\.[a-z]\+\.f32' /tmp/hipSYCL-*/nvptx-*.hipsycl_kernels
# It produces sqrt.rn.f32
```

Behavior does is the same with `-O0`.

**Expected behavior**

I would expect `sycl::native::sqrt` to result in the approximate version and `sycl::sqrt` in the IEE754-compliant one.

**Describe your setup**
* hipSYCL commit c0769a4a3da51caf821bddd723665211b7aa460d
* Describe the dependencies that hipSYCL sits on top of in your setup: 
   * Clang: 11.0.0-2~ubuntu20.04.1, installed from default Ubuntu 20.04 repository.
   * Same behavior with CUDA V11.2.142 and V11.0.221.
      * Files `/usr/local/cuda-11.x/version.txt` with the corresponding version strings were created to help Clang detect the CUDA versions.
* How you have compiled your application and which arguments you have passed to `syclcc`. In particular, which backends and hardware you have compiled for.
   * See above.

**Optional additional diagnostic information**
* For compilation/toolchain or setup-related issues: Setting the environment variable `HIPSYCL_DRYRUN=1` during compilation will cause `syclcc` to print the commands it would execute instead of actually executing. This can help verify the sanity of your setup.

```
/usr/bin/clang++-11 -I/home/aland/local/bin/../include/ -D__HIPSYCL__ -std=c++17 -U__FLOAT128__ -U__SIZEOF_FLOAT128__ -x cuda -D__HIPSYCL_ENABLE_CUDA_TARGET__ -D__HIPSYCL_CLANG__ -fplugin=/home/aland/local/bin/../lib/libhipSYCL_clang.so --cuda-path=/usr/local/cuda-11.2 --cuda-device-only --cuda-gpu-arch=sm_35 -S -o /tmp/hipSYCL-syclcc-8ffpu0vh/nvptx-sm_35.hipsycl_kernels -Wno-unknown-cuda-version -O3 test_sqrt.cpp
/usr/bin/clang++-11 -I/home/aland/local/bin/../include/ -D__HIPSYCL__ -std=c++17 -U__FLOAT128__ -U__SIZEOF_FLOAT128__ -x cuda -D__HIPSYCL_ENABLE_CUDA_TARGET__ -D__HIPSYCL_CLANG__ -fplugin=/home/aland/local/bin/../lib/libhipSYCL_clang.so --cuda-path=/usr/local/cuda-11.2 --cuda-device-only --cuda-gpu-arch=sm_75 -S -o /tmp/hipSYCL-syclcc-8ffpu0vh/nvptx-sm_75.hipsycl_kernels -Wno-unknown-cuda-version -O3 test_sqrt.cpp
/usr/bin/clang++-11 -I/home/aland/local/bin/../include/ -D__HIPSYCL__ -std=c++17 -D__HIPSYCL_MULTIPASS_CUDA_HEADER__="/tmp/hipSYCL-syclcc-8ffpu0vh/__hipsycl_embedded_cuda_kernels.hpp" -D__HIPSYCL_ENABLE_CUDA_TARGET__ -D__HIPSYCL_ENABLE_OMPHOST_TARGET__ -I/usr/include -Wno-unknown-cuda-version -O3 test_sqrt.cpp -o test -L/home/aland/local/bin/../lib/ -lhipSYCL-rt -Wl,-rpath=/home/aland/local/bin/../lib/ -Wl,-rpath=/usr/local/cuda-11.2/lib64 -L/usr/local/cuda-11.2/lib64 -lcudart -L/lib -lboost_context -lboost_fiber -lomp -Wl,-rpath=/lib
```

**Additional context**

It seems that [the explicitly chosen `__fsqrt_rn` intrinsic](https://github.com/illuhad/hipSYCL/blob/c0769a4a3da51caf821bddd723665211b7aa460d/include/hipSYCL/sycl/libkernel/math.hpp#L287) does what it's supposed to do. But Clang is pretty aggressive with its optimizations, so the "naive" `sqrt` becomes more optimized.

Since the [description of `__fsqrt_r*` intrinsics](https://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH__INTRINSIC__SINGLE.html#group__CUDA__MATH__INTRINSIC__SINGLE_1gf021e85b5e9de141a0fc2ff6fbe85875) do not mention that they are fast (unlike, e.g., [`__log2f`](https://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH__INTRINSIC__SINGLE.html#group__CUDA__MATH__INTRINSIC__SINGLE_1gfcc053f9040d50975aab00e44e7c6093)), it might be better to make `sycl::native::sqrt` use the plain `sqrt` and not intrinsics?

**Comment by illuhad** [comment]
Hm, interesting.

IMO the real bug here is that `sqrt()` is not IEEE754 compliant - the spec does not guarantee that the `native` functions *have* to be faster (they *may*), but of course it makes sense to expose approximating function variants when they exist.

AFAIK the default CUDA behavior should be to generate IEE754-compliant code, and indeed both nvcc and clang in godbolt seem to do this for `sqrt()`:
https://godbolt.org/z/cforMc

The question is why this doesn't happen for you.

Before we change anything we should better understand what is causing this and also check what happens on the HIP side.

Note: the `native` interface in hipSYCL is only partially implemented and might not work on the CPU backend.

**Comment by al42and** [comment]
> IMO the real bug here is that sqrt() is not IEEE754 compliant - the spec does not guarantee that the native functions have to be faster (they may), but of course it makes sense to expose approximating function variants when they exist.

Agree. I was approaching it from the optimization side, while accuracy is a more important problem in the grand scheme of things.

> AFAIK the default CUDA behavior should be to generate IEE754-compliant code, and indeed both nvcc and clang in godbolt seem to do this for sqrt():

That's an interesting observation.

Did more testing using your example, looks like Clang with CUDA-11.1 and newer are affected:

```
# Same output (with different register numbers) for clang-{10,11} -O{0,2}  --cuda-path=/usr/local/cuda-11.{1,2}:
$ clang-10 -O0 --cuda-path=/usr/local/cuda-11.2 --cuda-gpu-arch=sm_75 -Wno-unknown-cuda-version -std=c++17 --cuda-device-only -x cuda -S test_sqrt.cu -o - | grep sqrt 
        sqrt.approx.f32         %f4, %f3;

# Same output (with different register numbers) for clang-{10,11} -O{0,2}  --cuda-path=/usr/local/cuda-11.0:
$ clang-10 -O0 --cuda-path=/usr/local/cuda-11.0 --cuda-gpu-arch=sm_75 -Wno-unknown-cuda-version -std=c++17 --cuda-device-only -x cuda -S test_sqrt.cu -o - | grep sqrt 
        sqrt.rn.f32     %f4, %f3;


# Same output for /usr/local/cuda-11.{0,1,2}/bin/nvcc -O{0,2}
$ /usr/local/cuda-11.2/bin/nvcc -O0 -arch=sm_75 -c test_sqrt.cu -o /tmp/test_sqrt.o | cuobjdump /tmp/test_sqrt.o --extract-ptx kernel.1.sm_75.ptx && grep sqrt kernel.1.sm_75.ptx
sqrt.rn.f32 %f2, %f1;
```

> Note: the native interface in hipSYCL is only partially implemented and might not work on the CPU backend.

Yes, thanks for the warning. I can submit a quick fix to use fallback functions on CPU, if there's interest. It won't be fast but it will compile.

**Comment by illuhad** [comment]
We have since done a major refactoring/reimplementation of the builtin library. Brief update on the current behavior:
* The full native and half_precision interface is now implemented on all backends including host, but may not necessarily be mapped to actual native or half precision calls.
* Currently both native and non-native `sqrt()` variants are mapped to the regular `sqrt()`, so there should be no discrepancy anymore
* Why clang compiles to approximate functions I still don't know, but we should verify if the behavior persists with newer clang versions.
